linux/src/threadpool_linux.c

```diff
@@ -0,0 +1,322 @@
+                {
+                    // Timed out, loop through and try again
+                }
+                else
+                {
+                    // There was a new item added to the list, let's go
+                    break;
+                }
+            }
+            (void)interlocked_exchange(&threadpool->pool_state, POOL_STATE_ACTIVE);
+
+            // Locking
+            srw_lock_acquire_exclusive(threadpool->srw_lock);
+            if (threadpool->task_list != NULL && interlocked_add(&threadpool->quitting, 0) != 1)
+            {
+                // Remove the task item from the list
+                THREADPOOL_TASK* temp_item = threadpool->task_list;
+                threadpool->task_list = temp_item->next;
+                (void)interlocked_decrement(&threadpool->task_count);
+
+                srw_lock_release_exclusive(threadpool->srw_lock);
+
+                temp_item->task_func(temp_item->task_param);
+                free(temp_item);
+
+                srw_lock_acquire_exclusive(threadpool->srw_lock);
+            }
+            // Unlocking
+            srw_lock_release_exclusive(threadpool->srw_lock);
+
+        } while(interlocked_add(&threadpool->quitting, 0) != 1);
+        (void)interlocked_decrement(&threadpool->thread_count);
+    }
+    return 0;
+}
+
+THREADPOOL_HANDLE threadpool_create(EXECUTION_ENGINE_HANDLE execution_engine)
+{
+    THREADPOOL* result;
+    if (execution_engine == 0)
+    {
+        LogError("Invalid arguments: EXECUTION_ENGINE_HANDLE execution_engine: %p", execution_engine);
+        result = NULL;
+    }
+    else
+    {
+        result = malloc(sizeof(THREADPOOL));
+        if (result == NULL)
+        {
+            LogError("Failure allocating THREADPOOL structure");
+        }
+        else
+        {
+            const EXECUTION_ENGINE_PARAMETERS_LINUX* param = execution_engine_linux_get_parameters(execution_engine);
+
+            result->max_thread_count = param->max_thread_count;
+            if (result->max_thread_count == 0)
+            {
+                result->max_thread_count = MAX_THREAD_COUNT;
+            }
+
+            // Create a list of threads
+            result->thread_handle_list = malloc(sizeof(THREAD_HANDLE)*result->max_thread_count);
+            if (result->thread_handle_list == NULL)
+            {
+                LogError("Failure allocating THREAD array structure %zu", sizeof(THREAD_HANDLE)*result->max_thread_count);
+            }
+            else
+            {
+                result->list_index = 0;
+                result->srw_lock = srw_lock_create(false, "");
+                if (result->srw_lock == NULL)
+                {
+                    LogError("Failure srw_lock_create");
+                }
+                else
+                {
+                    result->thread_count = 0;
+                    result->task_count = 0;
+                    result->quitting = 0;
```

Comment:
>quitting [](http://example.com/codeflow?start=28&length=8)

interlocked

Reply:
I was wondering about this, since this is the only thread should I just = to zero or should I interlock it?

Reply:
We always access/modify volatile variables with interlocked.

There's a tiny chance that another thread starts using this after it has been created and that other thread did not execute any other instructions that created a barrier so it could see the stale data before it was initialized.

It probably wouldn't happen because some other interlocked call (e.g. the next line) would create the barrier, but now you are relying on weird side-effects.